SU2_CFD/include/solvers/CScalarSolver.hpp

```diff
@@ -0,0 +1,294 @@
+   * \param[in] solver_container - Container vector with all the solutions.
+   * \param[in] numerics - Description of the numerical method.
+   * \param[in] config - Definition of the particular problem.
+   */
+  template <class SolverSpecificNumericsFunc>
+  FORCEINLINE void Viscous_Residual_impl(const SolverSpecificNumericsFunc& SolverSpecificNumerics, unsigned long iEdge,
+                                         CGeometry* geometry, CSolver** solver_container, CNumerics* numerics,
+                                         CConfig* config) {
+    const bool implicit = (config->GetKind_TimeIntScheme() == EULER_IMPLICIT);
+    CVariable* flowNodes = solver_container[FLOW_SOL]->GetNodes();
+
+    /*--- Points in edge ---*/
+
+    auto iPoint = geometry->edges->GetNode(iEdge, 0);
+    auto jPoint = geometry->edges->GetNode(iEdge, 1);
+
+    /*--- Points coordinates, and normal vector ---*/
+
+    numerics->SetCoord(geometry->nodes->GetCoord(iPoint), geometry->nodes->GetCoord(jPoint));
+    numerics->SetNormal(geometry->edges->GetNormal(iEdge));
+
+    /*--- Conservative variables w/o reconstruction ---*/
+
+    numerics->SetPrimitive(flowNodes->GetPrimitive(iPoint), flowNodes->GetPrimitive(jPoint));
+
+    /*--- Turbulent variables w/o reconstruction, and its gradients ---*/
+
+    numerics->SetScalarVar(nodes->GetSolution(iPoint), nodes->GetSolution(jPoint));
+    numerics->SetScalarVarGradient(nodes->GetGradient(iPoint), nodes->GetGradient(jPoint));
+
+    /*--- Call Numerics contribution which are Solver-Specifc. Implemented in the caller: Viscous_Residual.  ---*/
+
+    SolverSpecificNumerics(*geometry, *numerics, *config, *nodes, iPoint, jPoint);
+
+    /*--- Compute residual, and Jacobians ---*/
+
+    auto residual = numerics->ComputeResidual(config);
+
+    if (ReducerStrategy) {
+      EdgeFluxes.SubtractBlock(iEdge, residual);
+      if (implicit) Jacobian.UpdateBlocksSub(iEdge, residual.jacobian_i, residual.jacobian_j);
+    } else {
+      LinSysRes.SubtractBlock(iPoint, residual);
+      LinSysRes.AddBlock(jPoint, residual);
+      if (implicit) Jacobian.UpdateBlocksSub(iEdge, iPoint, jPoint, residual.jacobian_i, residual.jacobian_j);
+    }
+  }
+
+ private:
+  /*!
+   * \brief Compute the viscous flux for the turbulent equation at a particular edge.
+   * \param[in] iEdge - Edge for which we want to compute the flux
+   * \param[in] geometry - Geometrical definition of the problem.
+   * \param[in] solver_container - Container vector with all the solutions.
+   * \param[in] numerics - Description of the numerical method.
+   * \param[in] config - Definition of the particular problem.
+   * \note Calls a generic implementation after defining a SolverSpecificNumerics object.
+   */
+  inline virtual void Viscous_Residual(unsigned long iEdge, CGeometry* geometry, CSolver** solver_container,
+                                       CNumerics* numerics, CConfig* config) {}
```

Comment:
You can give it default behavior by passing an empty solver-specific object.
If you see strong reasons against that, make this pure virtual so that no one forgets this needs to be implemented in derived classes.

Reply:
I went with the empty solver specific implementation instead of pure virtual. I guess as of right now all solvers will need to their own stuff so one could pure virtual imo but maybe there will be a gneric Diffusivity at one point and then this might come in handy
Done in https://github.com/su2code/SU2/pull/1330/commits/1d8f520a9da2907b105b3f47f2fcd05e75d424c9